### ISSUE TITLE
Names and terms used in egor

### ISSUE THREAD

**Issue by tilltnet** [tilltnet/egor#8]
For egonetR we arrived at the consensus of the following set of variable names that egonetR would use as default values and in the documentation.

**IDs**
egoID
alterID

**df names**
egos
alteri (changed to alters in egor since it is more common in english)
edges (in egor currently alter_ties)

**edge list**
Source
Target
(weight)

I think the biggest problem are the alter-alter relations. Edges and ties are sometimes used in the literature to describe the links from egos to the alters as well as for the relations between the alters. I am failing to come up with a good solution. My ideas are 'aa_ties' and 'alter_alter'. Any ideas are welcome!

**Comment by krivit** [comment]
@martinamorris, this may be of interested to you as well.

I am increasingly liking `.aa_ties` or even `.aaties` to save a shift keystroke.

In fact, I wonder if we should just use `.a` for alters table and `.aa` for the alter-alter table, and generally shorten frequently-referenced columns as much as possible:
```
subset(egor, abs(.alters$age[.alter_ties$Source] -
             .alters$age[.alter_ties$Target])<=5,
       aspect = "ties")
```
becomes (using shortcuts in https://github.com/tilltnet/egor/issues/7#issuecomment-302953289)
```
subset(egor, abs(.a$age[.aa$.src] -
             .a$age[.aa$.tgt])<=5,
       aspect = "ties")
```

**Comment by krivit** [comment]
Also, I suggest prepending all "reserved" column names with dots. So, if we have a standard alter ID, we should call the column .alterID and the alter-alter matrix's columns .source and .target. This way, if the user actually collected an alter-alter relation attribute named "Source" or "Target", they won't be clobbering anything.

**Comment by krivit** [comment]
@tilltnet, what do you think about the "dot policy"?

**Comment by krivit** [comment]
In UW NMG group, the consensus seemed to be that `.alters` was OK as is, but `.alter_ties` might be worth renaming to `.aaties` to disambiguate and save typing.

Some weren't happy about the "dot policy".

**Comment by tilltnet** [comment]
.aaties would be fine with me as well.
 
What are the arguments against the 'dot policy'? My personal first thought was that it would look a bit more cryptic, but I feel that having names, that most certainly won't collide with names used by the users outweighs this.

**Comment by krivit** [comment]
Picking up from #7, @tilltnet, I guess that leaves the question of `Source` and `Target` columns. I like the idea of dot-lowercasing them (i.e., `.source` and `.target` or some other standard names), but that's something affecting the import utilities, which is something I know next to nothing about.

Personally, I prefer something neutral like `.a1` and `.a2`, or `.alter1` and `.alter2`, especially since most of these relations will be undirected.

**Comment by martinamorris** [comment]
On Thu, 1 Jun 2017, Pavel N. Krivitsky wrote:

> 
> Picking up from #7, @tilltnet, I guess that leaves the question of Source and Target columns. I like the idea of
> dot-lowercasing them (i.e., .source and .target or some other standard names), but that's something affecting the
> import utilities, which is something I know next to nothing about.
> 
> Personally, I prefer something neutral like .a1 and .a2, or .alter1 and .alter2, especially since most of these
> relations will be undirected.

i like .a1 and .a2

if we're going to use the "a" abbreviation for aaties, then we may as well 
go all in and use "a" here (and other places)

**Comment by krivit** [comment]
@tilltnet , @raffaelevacca , @martinamorris , can we make finalizing this a priority? Once this issue and #7 are resolved, we can start importing real-world datasets into `egor`, reasonably confident that the data format isn't going to change.

I hereby volunteer to go through the code and make the agreed-upon renames where needed.

**Comment by martinamorris** [comment]
can you summarize the items that need to be named, and the winnowed out suggestions at this point?

**Comment by krivit** [comment]
> can you summarize the items that need to be named, and the winnowed out suggestions at this point?

I'm putting together a wiki page to do just that right now.

**Comment by krivit** [comment]
@tilltnet @raffaelevacca @martinamorris 
I've created a [wiki page](https://github.com/tilltnet/egor/wiki/egor-format-specification) that summarizes the format *as implemented* (at least as I understand it). Feel free to edit.

The page also lists pending proposals to amend this specification, which should hopefully allow us to organize the discussion and actually come to some decisions. :-)

**Comment by raffaelevacca** [comment]
Hi @krivit and All,
I've gone through the wiki page and all linked discussions. These are my two cents on the amendment proposals:
1. OK to replace `.*Ix` with `.*Row`.
2. I would adopt **Proposal 3** in https://github.com/tilltnet/egor/wiki/egor-format-specification#2-alter-id-column-fixed-or-variable. I think `.aaties$Source` and `.aaties$Target` should reference a fixed alter ID column, which I would call `.alterID` or (shortened version) `.altID`. I understand the advantages of proposal 1, but I agree with @tilltnet that that option could be "too advanced" for the average user and create confusion and doubts about data integrity. Proposal 3 is more "old school" and probably more familiar to users. Although it wouldn't be strictly needed in `egor`, I believe it would help average users to better understand how `egor` works and to have confidence that the package is doing what they think it's doing.
3. I agree that special-purpose names should be standardized to begin with a dot (`.`).
4. I agree with most renaming proposals. Clearly it's a trade-off between short/simple shortcuts and shortcuts that still convey some meaning to the user. I suspect that just an `a` would not immediately mean `alter` to most users. So here's my two cents:
    * `.alters` -> `.alts`
    * `.aaties` -> `.aat`
    * `alterID` -> `.altID`
    * `Source` -> `.alt1`
    * `Target` -> `alt2`
I appreciate all the progress you're making, thank you for all your work!

**Comment by tilltnet** [comment]
On '**2.** Alter ID column: fixed or variable', i think the third option with a fixed alterID column is the way to go.

> A sub-question is whether or not, if proposals 2 or 3 get adopted, the code should fall back to 1 if the column in question is absent.

If falling back it should be done with a warning. Maybe a stop/error is even better in order to enourage the user to make sure the data matches up?!

On '**3.** dot policy': until I hear a good argument against it I am in favor of it.

On '**4.** name changes' I like what @raffaelevacca said. Single letter variable names are a bit too much I think. But I would prefer srcID and tgtID. Because in case of the alter-alter relations being directed this would make it obvious what the direction is.

**Comment by krivit** [comment]
My thoughts so far:

A. I think that `.aaties` is already short enough and doesn't require any SHIFT presses, and `.aat` is less descriptive.
B. If we shorten `.alters` to `.alts`, should we do that everywhere? For example, we have function arguments like `max.alters`.
C. I like the idea of using just `.alts$.alt` as the alter ID, and then using `.aaties$.alt1` and `.aaties$.alt2` as the source and target column, since the direction (if any) is obvious.

@martinamorris, any thoughts?

**Comment by krivit** [comment]
@tilltnet, @raffaelevacca, @martinamorris, I'd like to make one more push to finalize this.

It looks like the main points of contention are:

1. `.aaties` vs. `.aat`
2. `.altID` vs. `.alt`
3. `.srcID`/`.tgtID` vs. `.alt1`/`.alt2`

My own preferences are `.aaties`, `.alt`, and `.alt1`/`.alt2`, the last two as a package. (I.e., if one has an "ID" in it, so should the other.) What do y'all think?

**Comment by martinamorris** [comment]
On Fri, 9 Jun 2017, Pavel N. Krivitsky wrote:

> 
> @tilltnet, @raffaelevacca, @martinamorris, I'd like to make one more push to finalize this.
> 
> It looks like the main points of contention are:
>
>  1. .aaties vs. .aat
>  2. .altID vs. .alt
>  3. .srcID/.tgtID vs. .alt1/.alt2
> 
> My own preferences are .aaties, .alt, and .alt1/.alt2, the last two as a package. (I.e., if one has an "ID" in it, so
> should the other.) What do y'all think?

1. .aaties is fine, but i'm happy with either
2. .altID (for clarity) or .aID (see below)

3. it's a bit unclear the 1 and 2 refer to, and at the risk of 
muddying the water with more options, what about

.a1ID/.a2ID

**Comment by krivit** [comment]
Well, it's extra shift-keystrokes, but virtual columns have those as well.

So, then, `.altID` and `.alt1ID` and `.alt2ID`?

Should it be `.alt1ID` or `.altID1`? Any thoughts? @tilltnet @martinamorris @raffaelevacca ?

**Comment by martinamorris** [comment]
On Sat, 10 Jun 2017, Pavel N. Krivitsky wrote:

> 
> Well, it's extra shift-keystrokes, but virtual columns have those as well.
> 
> So, then, .altID and .alt1ID and .alt2ID?
> 
> Should it be .alt1ID or .altID1? Any thoughts? @tilltnet @martinamorris @raffaelevacca ?

actually, yes, .altID1 and .altID2 are better


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AH0J4mZymBJBWJfcgtmWPUt5sOySwCcSks5sCv3RgaJpZM4NhktF.gif]
> 
> 
>

****************************************************************
  Professor of Sociology and Statistics
  Director, UWCFAR Sociobehavioral and Prevention Research Core
  Box 354322
  University of Washington
  Seattle, WA 98195-4322

  Office:        (206) 685-3402
  Dept Office:   (206) 543-5882, 543-7237
  Fax:           (206) 685-7419

morrism@u.washington.edu
http://faculty.washington.edu/morrism/

**Comment by tilltnet** [comment]
I don't like numbers in variable names since they make the name look temporary to me. Therefore I suggest we use:

- .aaties
- .altID
- .srcID/ .tgtID

**Comment by martinamorris** [comment]
> I don't like numbers in variable names since they make the name look temporary to me. Therefore I suggest we use:
>
>  *  .aaties
>  *  .altID
>  *  .srcID/ .tgtID

i think the issue we had with src/tgt is that it implies directed ties, 
and in the undirected context, will lead to confusion.

the numbered versions (altID1/altID2) can be mapped to both contexts: in 
undirected, the mapping would be in terms of ID values (arbitrary, but 
consistent); in directed it would be in terms of src/tgt.

we currently use b1/b2 for the 2 modes in a bipartite network in ergm.  so 
there's that...

and then there's the possibility of allowing both for user syntax, and 
translating to whatever we like inside the function.

**Comment by tilltnet** [comment]
> i think the issue we had with src/tgt is that it implies directed ties, and in the undirected context, will lead to confusion.

As stated above, I see this ambiguity as a feature. Data might include directed or undirected alter-alter ties, in case of undirected data, the implied direction can simply be ignored. If the aaties are directed the names src/tgt give a clear indication of the direction. With alt1 and alt2 the direction could also be interpreted as 1 receives from 2.

**Comment by krivit** [comment]
@martinamorris, @tilltnet also worth keeping in mind that the data might have reports for both directed and undirected alter-alter relations.

Also, for undirected relations, should we specify that it should be listed twice (i.e., once from each end) or once? If once, should we allow any direction, or mandate that, say, `.srcID < .dstID`?

**Comment by martinamorris** [comment]
> @martinamorris, @tilltnet also worth keeping in mind that the data might have reports for both directed and
> undirected alter-alter relations.

ok. but any specific relation must be one or the other, yes?  so you'd 
only have the possibility of both types in the case of multiplex ties.

is there any value in using different syntax to refer directed vs.
undirected ties?

how is this handled for ego-alter ties?  those can also be directed or 
undirected.  in a sense ego always "nominates" alters, but the nomination 
could be "received advice from" or "gave advice to".

> Also, for undirected relations, should we specify that it should be listed twice (i.e., once from each end) or once?
> If once, should we allow any direction, or mandate that, say, .srcID < .dstID?

ugh to listed twice.

iirc, the ordering issue also arises in network/ergm with undirected ties. 
i don't recall the specifics, but we used the head:tail nominclature 
there, and i think a convention of headID < tailID (or the reverse) for 
undirected data.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AH0J4nuAb2Y-8DohsIOwvE-En4WO38GOks5sDD6WgaJpZM4NhktF.gif]
> 
> 
>

****************************************************************
  Professor of Sociology and Statistics
  Director, UWCFAR Sociobehavioral and Prevention Research Core
  Box 354322
  University of Washington
  Seattle, WA 98195-4322

  Office:        (206) 685-3402
  Dept Office:   (206) 543-5882, 543-7237
  Fax:           (206) 685-7419

morrism@u.washington.edu
http://faculty.washington.edu/morrism/

**Comment by krivit** [comment]
We decided to table the directedness issue until we have examples of data and applications for it.

>  iirc, the ordering issue also arises in network/ergm with undirected ties. i don't recall the specifics, but we used the head:tail nominclature there, and i think a convention of headID < tailID (or the reverse) for undirected data.

It's `tailID<headID` for undirected networks.

**Comment by krivit** [comment]
Upon further thought, I don't find a direction in column name *that* objectionable. In the `ergm` package, we talk about tails and heads, even for undirected networks, for example.

Either `.srcID`/`.tgtID` or `.tailID`/`.headID` are fine by me. The latter has an advantage of being more neutral, but the former is shorter. @martinamorris, @tilltnet , @raffaelevacca , any thoughts? I think this might be the last question before we can "freeze" the format.

**Comment by martinamorris** [comment]
> Either .srcID/.tgtID or .tailID/.headID are fine by me. The latter has an advantage of being more neutral, but the
> former is shorter. @martinamorris, @tilltnet , @raffaelevacca , any thoughts? I think this might be the last question
> before we can "freeze" the format.

slight pref for consistency with ergm

**Comment by tilltnet** [comment]
I think source and target (src/tgt) are common column names for edgelists. But tail and head are fine as well, I  have to say that it doesn't feel as intuitive to me, as source and target, but that might be due to me not being a native english speaker. @andreasys @raffaelevacca I'd really like to hear your opinions on this one!

**Comment by andreasys** [comment]
First of all thanks for all you work! 
I have a slight preference for .srcID/.tgtID as it expresses the directedness "better" than the .tailID/.headID. Though that might be because i only worked with source/ target column names so far..

**Comment by raffaelevacca** [comment]
Honestly I have no preference on this one, there are good reasons for either choice. So feel free to make the call as far as I'm concerned. Thank you!

**Comment by martinamorris** [comment]
> 
> First of all thanks for all you work!
> I have a slight preference for .srcID/.tgtID as it expresses the directedness "better" than the .tailID/.headID.
> Though that might be because i only worked with source/ target column names so far..

i agree src/tgt is clearer.  probably the better choice then.

****************************************************************
  Professor of Sociology and Statistics
  Director, UWCFAR Sociobehavioral and Prevention Research Core
  Box 354322
  University of Washington
  Seattle, WA 98195-4322

  Office:        (206) 685-3402
  Dept Office:   (206) 543-5882, 543-7237
  Fax:           (206) 685-7419

morrism@u.washington.edu
http://faculty.washington.edu/morrism/

**Comment by krivit** [comment]
Looks like we have a consensus. I've updated the wiki.

**Comment by krivit** [comment]
@martinamorris, @tilltnet , @raffaelevacca, @andreasys , I think this is all the questions we had for the format. Is there anything else we need to discuss on this issue?
